app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt

```diff
@@ -7,90 +7,90 @@ import de.westnordost.streetcomplete.ktx.format
+import kotlin.coroutines.resumeWithException
 
 class MapTilesDownloader @Inject constructor(
     private val vectorTileProvider: VectorTileProvider,
     private val cacheConfig: MapTilesDownloadCacheConfig
 ) {
     private val okHttpClient = OkHttpClient.Builder().cache(cacheConfig.cache).build()
 
-    data class Tile(val zoom: Int, val x: Int, val y: Int)
-
     suspend fun download(bbox: BoundingBox) = withContext(Dispatchers.IO) {
         var tileCount = 0
         var failureCount = 0
         var downloadedSize = 0
         var cachedSize = 0
         val time = currentTimeMillis()
 
-        coroutineScope {
-            for (tile in getDownloadTileSequence(bbox)) {
-                launch {
-                    val result = try {
-                        downloadTile(tile.zoom, tile.x, tile.y)
-                    } catch (e: Exception) {
-                        DownloadFailure
-                    }
-                    ++tileCount
-                    when (result) {
-                        is DownloadFailure -> ++failureCount
-                        is DownloadSuccess -> {
-                            if (result.alreadyCached) cachedSize += result.size
-                            else downloadedSize += result.size
-                        }
-                    }
+        downloadTilesFlow(bbox).collect { result ->
+            ++tileCount
+            when (result) {
+                is DownloadFailure -> ++failureCount
+                is DownloadSuccess -> {
+                    if (result.alreadyCached) cachedSize += result.size
+                    else downloadedSize += result.size
                 }
             }
         }
+
         val seconds = (currentTimeMillis() - time) / 1000.0
         val failureText = if (failureCount > 0) ". $failureCount tiles failed to download" else ""
         Log.i(TAG, "Downloaded $tileCount tiles (${downloadedSize / 1000}kB downloaded, ${cachedSize / 1000}kB already cached) in ${seconds.format(1)}s$failureText")
     }
 
-    private fun getDownloadTileSequence(bbox: BoundingBox): Sequence<Tile> =
+    private fun downloadTilesFlow(bbox: BoundingBox): Flow<DownloadResult> = channelFlow {
         /* tiles for the highest zoom (=likely current or near current zoom) first,
            because those are the tiles that are likely to be useful first */
-        (vectorTileProvider.maxZoom downTo 0).asSequence().flatMap { zoom ->
-            bbox.enclosingTilesRect(zoom).asTilePosSequence().map { Tile(zoom, it.x, it.y) }
+        (vectorTileProvider.maxZoom downTo 0).forEach { zoom ->
+            bbox.enclosingTilesRect(zoom).asTilePosSequence().forEach { (x, y) ->
+                launch {
+                    try {
+                        val result = downloadTile(zoom, x, y)
+                        val logText = if (result.alreadyCached) "in cache" else "downloaded"
+                        Log.v(TAG, "Tile $zoom/$x/$y $logText")
+                        send(result)
```

Comment:
The `send` call here produces a new warning about not being marked with the `@ExperimentalCoroutinesApi` decorator, which I believe is incorrect.

[`channelFlow`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/channel-flow.html) is not experimental. However, the [`ProducerScope`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.channels/-producer-scope/index.html) block that you pass is. As far as I can tell, it's only marked that way because a couple of its functions are experimental — *not* including [`send`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.channels/-send-channel/send.html) — but the compiler isn't smart enough to tell that the part of the interface we're using isn't actually experimental.

Reply:
I suppose I should probably include a comment to that effect.